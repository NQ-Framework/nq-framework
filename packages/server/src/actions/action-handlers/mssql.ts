import {
  ActionHandler,
  ActionInstance,
  PropertyValue,
  WorkflowExecutionContext,
} from '@nqframework/models';
import { ModuleRef } from '@nestjs/core';
import { RequestRouterService } from '../../db-connection/request-router/request-router.service';
import { MsSqlFetcher } from '@nqframework/data-fetcher';

export const handler: ActionHandler = {
  handle: async (
    propertyValues: PropertyValue[],
    actionInstance: ActionInstance,
    workflowExecution: WorkflowExecutionContext,
    moduleRef: ModuleRef,
  ): Promise<PropertyValue[]> => {
    const query = propertyValues.find((i) => i.name === 'query')?.value;
    const credentials = propertyValues.find((i) => i.name === 'credentials')
      ?.value;
    const userId = propertyValues.find((i) => i.name === 'userId')?.value;
    const isProcedure = propertyValues.find((i) => i.name === 'isProcedure')?.value;
    if (!query) {
      throw new Error('Missing required parameter query');
    }
    if (!credentials) {
      throw new Error('Missing required parameter credentials');
    }
    if (!userId) {
      throw new Error('Missing required parameter userId');
    }
    if (isProcedure === undefined) {
      throw new Error('Missing required parameter isProcedure');
    }

    const requestRouter = moduleRef.get<RequestRouterService>(
      RequestRouterService,
      { strict: false },
    );

    const fetcher: MsSqlFetcher = (await requestRouter.getDataFetcher(
      userId,
      workflowExecution.workflow.organizationId,
      credentials,
    )) as any;
    const result = await fetcher.get({
      isProcedure,
      query,
    });

    return [
      {
        name: 'queryResult',
        value: result,
      },
    ];
  },
};
