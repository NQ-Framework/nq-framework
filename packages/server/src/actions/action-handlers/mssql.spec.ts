import { handler } from './mssql';

describe('mssql action', () => {
  it('should execute query and return result', async () => {
    const result = await handler.handle(
      [
        { name: 'query', value: 'mock query' },
        { name: 'credentials', value: 'credentials' },
        { name: 'userId', value: 'mock user id' },
      ],
      {} as any,
      { workflow: { organizationId: 'mock org id' } } as any,
      {
        get: jest.fn().mockImplementation(() => ({
          getDataFetcher: () => ({
            get: () => ({
              rows: [{ testCol: 'test val' }],
            }),
          }),
        })),
      },
    );
    expect(result).toEqual([
      { name: 'queryResult', value: { rows: [{ testCol: 'test val' }] } },
    ]);
  });
  it('should throw when required params are not present', async () => {
    await expect(
      handler.handle(
        [{ name: 'non query', value: 'test' }],
        {} as any,
        {} as any,
        { get: jest.fn() } as any,
      ),
    ).rejects.toThrowErrorMatchingSnapshot();
    await expect(
      handler.handle(
        [{ name: 'query', value: 'test' }],
        {} as any,
        {} as any,
        { get: jest.fn() } as any,
      ),
    ).rejects.toThrowErrorMatchingSnapshot();
    await expect(
      handler.handle(
        [
          { name: 'query', value: 'test' },
          { name: 'credentials', value: 'test' },
        ],
        {} as any,
        {} as any,
        { get: jest.fn() } as any,
      ),
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});
