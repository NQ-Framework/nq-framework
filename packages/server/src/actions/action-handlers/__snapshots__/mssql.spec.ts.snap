// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mssql action should throw when required params are not present 1`] = `"Missing required parameter query"`;

exports[`mssql action should throw when required params are not present 2`] = `"Missing required parameter credentials"`;

exports[`mssql action should throw when required params are not present 3`] = `"Missing required parameter userId"`;

exports[`mssql action should throw when required params are not present 4`] = `"Missing required parameter isProcedure"`;
